src/crimson/common/operation.h

```diff
@@ -525,15 +529,25 @@ class PipelineHandle {
   seastar::future<>
   enter(T &stage, typename T::BlockingEvent::template Trigger<OpT>&& t) {
     ceph_assert(stage.get_core() == seastar::this_shard_id());
-    return wait_barrier().then([this, &stage, t=std::move(t)] () mutable {
-      auto fut = t.maybe_record_blocking(stage.enter(t), stage);
-      exit();
-      return std::move(fut).then(
-        [this, t=std::move(t)](auto &&barrier_ref) mutable {
-        barrier = std::move(barrier_ref);
-        return seastar::now();
+    if (barrier && barrier->type == "Concurrent") {
```

Comment:
I don't think I understand what this is trying to fix.  The difference between these two cases is whether we invoke wait_barrier(), right?  If !barrier, wait_barrier() will be seastar::now().  OrderedExclusivePhaseT and UnorderedStageT have trivial barriers where wait() will return seastar::now().  Is returning seastar::now() like that causing control to return to the reactor?

Reply:
in debug build, if wait_barrier is seastar::now, will cause reactor schedule the continuation away,  but can't schedule them back in order by which they are scheduled away, so the continuations are out of order here.

Reply:
I don't like the type member/comparisons.  Instead, modify the wait_barrier interface and implementations to return std::optional<seastar::future<>> and wait only if non-empty.

Reply:
done